python/lance/types/box.py

```diff
@@ -30,84 +33,211 @@ def __init__(self):
             self.flatten()
             .to_numpy(zero_copy_only=zero_copy_only)
-            .reshape((len(self), 4))
+            .reshape((len(self), (self.ndims * 2)))
         )
 
+    def get_length(self, axis: Union[int, str]):
+        axis = self._sanitize_axis(axis)
+        return self.get_max(axis) - self.get_min(axis) + 1
+
+    def get_min(self, axis: Union[int, str]):
+        """
+        Get a 1-d numpy array of the minimums for each box along given axis
+
+        Parameters
+        ----------
+        axis: int or str
+            0 is 'x', 1 is 'y', 2 is 'z`
+        """
+        axis = self._sanitize_axis(axis)
+        return self.to_numpy()[:, axis]
+
+    def get_max(self, axis: Union[int, str]):
+        """
+        Get a 1-d numpy array of the maximums for each box along given axis
+
+        Parameters
+        ----------
+        axis: int or str
+            0 is 'x', 1 is 'y', 2 is 'z`
+        """
+        axis = self._sanitize_axis(axis)
+        return self.to_numpy()[:, self.ndims + axis]
+
+    def _sanitize_axis(self, axis):
+        if isinstance(axis, str):
+            axis = {"x": 0, "y": 1, "z": 2}[axis]
+        if axis < 0 or axis >= self.ndims:
+            raise ValueError(
+                f"Axis {axis} is greater the number of box " f"dimensions in this array"
+            )
+        return axis
+
     @property
     def xmin(self) -> np.ndarray:
         """Return a numpy array of the min X-coord of each box"""
-        return self.to_numpy()[:, 0]
+        return self.get_min("x")
 
     @property
     def ymin(self) -> np.ndarray:
         """Return a numpy array of the min Y-coord of each box"""
-        return self.to_numpy()[:, 1]
+        return self.get_min("y")
 
     @property
     def xmax(self) -> np.ndarray:
         """Return a numpy array of the max X-coord of each box"""
-        return self.to_numpy()[:, 2]
+        return self.get_max("x")
 
     @property
     def ymax(self) -> np.ndarray:
         """Return a numpy array of the max Y-coord of each box"""
-        return self.to_numpy()[:, 3]
+        return self.get_max("y")
+
+    def _box_sizes(self):
+        sizes = self.get_length(0)
+        for i in range(1, self.ndims):
+            sizes *= self.get_length(i)
+        return sizes
+
+    def iou(self, others: "BoxNdArray") -> np.ndarray:
+        """
+        Compute the intersection-over-union between these bounding boxes
+        and the given bounding boxes.
+
+        Return
+        ------
+        ious: an np.ndarray
+            If n = len(self), m = len(others) then returned ious should
+            have shape (n, m)
+        """
+        if not isinstance(others, BoxNdArray):
+            raise TypeError("Need BoxNDArray to compute IOU")
+        if self.ndims != others.ndims:
+            raise ValueError("ndims must be equal to compute IOU")
+        size_self = self._box_sizes()
+        if self is others:
+            size_others = size_self
```

Comment:
should we just return `np.array([1, 1, 1, 1, ...])` here?

Otherwise, this seems to just save one compute of box sizes? i feel that it is not necessary at the moment.

Reply:
it's all-pairs so even if we're doing box_arr.iou(box_arr), it should return a matrix of ious no?

Reply:
then it returns a `np.ones(shape=(self.array.length, self.array.length))`?

Reply:
Why would it return all ones? If you have a list of boxes, box_i.iou(box_j) won't always be 1?